BFloat16: enable prepacked weights's inference

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48994 enable mkldnn conv2d backward to support mkldnn tensor input
* **#48922 BFloat16: enable prepacked weights's inference**

Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

XiaobingSuper committed:
BFloat16: enable prepacked weights's inference

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"




[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin , the error is caused by runing old cpu device(only support avx2), but onednn need the device support avx512 at least for bfloat16 path, so I want to add a gloabal context like **torch._C.has_mkldnn** to skip bfloat16 test case if cpu device not support avx512,. is it ok for your side?

XiaobingSuper commented:
@VitalyFedyunin, There has another option: fall back fp32 path in ideep if onednn is not supported on some old device, this option don't need to change pytorch code except updata ideep. if this option is ok, I will updata ideep at first. Thanks!

jgong5 commented:
> There has another option: fall back fp32 path in ideep if onednn is not supported on some old device, this option don't need to change pytorch code except updata ideep. if this option is ok, I will updata ideep at first. Thanks!

@VitalyFedyunin This option sounds better to us since it does not require extra CPU arch check from the PyTorch side. IDEEP will support BF16 regardless of CPU archs. We will have a quick upgrade on IDEEP if it looks good to you. :-)

VitalyFedyunin commented:
Current error doesn't tell anything about avx support:

```
        x_mkldnn = x if x.is_mkldnn else x.to_mkldnn()
        y_mkldnn = torch._C._nn.mkldnn_linear(x_mkldnn, self.weight, self.bias)
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        y = y_mkldnn if x.is_mkldnn else y_mkldnn.to_dense()
```

Falling back to fp32 path will create invalid impression for users. It is much better to error out with nice warning when something is not supported. 

For tests you don't need to introduce new global context. Something simple like @skipIf(test_arch(),"not supported architecture") will suffice.

jgong5 commented:
> Falling back to fp32 path will create invalid impression for users. It is much better to error out with nice warning when something is not supported.

@VitalyFedyunin Thanks for the suggestions. So, we will also add cpu arch check inside `use_mkldnn` (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L227) so that it does not support bf16 without avx512. Does that make sense to you?

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

VitalyFedyunin commented:
Ok we fixed tests, but can we please make sure that we are going to have nice error if someone tries it on old CPU

XiaobingSuper commented:
> Ok we fixed tests, but can we please make sure that we are going to have nice error if someone tries it on old CPU

There will has a simple API to detect cpu device at next onednn upgrade,, and will report error if one cpu not support bfloat16.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper commented:
@VitalyFedyunin , please help review it. Thanks!

ngimel commented on test/test_mkldnn.py:
```diff
@@ -192,6 +203,30 @@ def test_conv2d(self):
             self._test_serialization(mkldnn_conv2d, (x.to_mkldnn(),))
             self._test_tracing(mkldnn_conv2d, (x.to_mkldnn(),))
 
+    @unittest.skipIf(not has_bf16_support(), "OneDNN bfloat16 path requires AVX512")
```
these tests should not be skipped if there's no bf16 support, instead they should be catching and asserting that appropriate errors are raised.

XiaobingSuper replied:
Add test cases if there's no bf16 support.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper commented on torch/utils/mkldnn.py:
```diff
@@ -178,25 +178,25 @@ def forward(self, x):
         if isinstance(m, torch.nn.Linear):
-            return MkldnnLinear(m)
+            return MkldnnLinear(m, d)
         elif isinstance(m, torch.nn.Conv1d):
```
Dtype argument is added on Conv1d, Conv3d and will assert if given dtype is not supported.

XiaobingSuper commented on test/test_mkldnn.py:
```diff
@@ -19,6 +19,17 @@
+    try:
+        # for bf16 path, OneDNN requires the cpu has intel avx512 with avx512bw,
+        # avx512vl, and avx512dq.
+        cmd = "grep avx512bw /proc/cpuinfo | grep avx512vl | grep avx512dq"
```
For windows, bf16 cases are skiped now, the reason is that I can't find a easy way to check the device as linux.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

ngimel commented on torch/utils/mkldnn.py:
```diff
@@ -178,25 +177,28 @@ def forward(self, x):
-            return MkldnnConv3d(m)
+            return MkldnnConv3d(m, d)
         elif isinstance(m, torch.nn.BatchNorm2d) or isinstance(m, torch.nn.BatchNorm3d):
             return MkldnnBatchNorm(m)
```
why is dtype ignored for batchnorm? Does it expect its parameters to remain in fp32?

XiaobingSuper replied:
Yes, batchnorm's parameters remain in fp32.

ngimel reviewed (commented):
This PR would really benefit from better description and documentation, e.g. the fact that bfloat16 in mkldnn is supported only on avx512 processors really should be put in the description and in the Note somewhere in the code. 
Similarly, I would like to understand why bias is never converted to bfloat16. If it is by design, it should be commented upon and put in a note. 
Similarly, for BatchNorm parameters - if OneDNN wants them in fp32, that's fine, but there should be a comment about this.

XiaobingSuper commented on torch/utils/mkldnn.py:
```diff
@@ -1,11 +1,10 @@
-        self.register_buffer('weight', dense_module.weight.to_mkldnn())
+        self.register_buffer('weight', dense_module.weight.to_mkldnn(dtype))
         if dense_module.bias is not None:
             self.register_buffer('bias', dense_module.bias.to_mkldnn())
```
Yes, for better accuracy, bias is expected fp32 type fro bfloat16 module.

XiaobingSuper commented on test/test_mkldnn.py:
```diff
@@ -144,53 +156,103 @@ def test_repr(self):
+
+    def test_conv1d(self):
+        for groups in [1, 4]:
+            N = torch.randint(3, 10, (1,)).item()
```
Changed.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper commented:
> This PR would really benefit from better description and documentation, e.g. the fact that bfloat16 in mkldnn is supported only on avx512 processors really should be put in the description and in the Note somewhere in the code.
> Similarly, I would like to understand why bias is never converted to bfloat16. If it is by design, it should be commented upon and put in a note.
> Similarly, for BatchNorm parameters - if OneDNN wants them in fp32, that's fine, but there should be a comment about this.

Yes, I add some comments in **test_mkldnn.py** and **torch/utils/mkldnn.py** to describe them.

XiaobingSuper commented on test/test_mkldnn.py:
```diff
@@ -250,24 +289,98 @@ def test_relu_(self):
+    @unittest.skipIf(IS_WINDOWS, "Limit support for bf16 path")
+    def test_relu_inplace_bf16(self):
+        x = torch.randn((4, 5), dtype=torch.float32) * 10
+        x_bf16 = x.clone().bfloat16()
```
Changed.

XiaobingSuper commented on test/test_mkldnn.py:
```diff
@@ -250,24 +289,98 @@ def test_relu_(self):
+                                   lambda: torch.relu(x_bf16.to_mkldnn()))
+
+    @unittest.skipIf(IS_WINDOWS, "Limit support for bf16 path")
+    def test_relu_inplace_bf16(self):
```
Thanks, I changed it.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper commented:
@ngimel , the failed case seems not related to this PR.

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "BFloat16: enable prepacked weights's inference"


Differential Revision: [D25537188](https://our.internmc.facebook.com/intern/diff/D25537188)

[ghstack-poisoned]
